src/engine/node.rs

```diff
@@ -0,0 +1,216 @@
+use std::fmt;
+
+use super::state::State;
+
+pub enum Op {
+    Add,
+    Mul,
+    Tanh,
+    Exp(f32),
+    Pass,
+    End,
+}
+
+pub struct Node {
+    pub data: f32,
+    pub name: String,
+    pub grad: f32,
+    pub _idx: usize,
+    pub _prev: (Option<usize>, Option<usize>),
+    pub _op: Op,
+    pub _parameter: bool,
+}
+
+impl Node {
+    pub fn new(data: f32, name: String, is_parameter: bool, state: &mut State) -> usize {
+        let out = Node {
+            data,
+            name,
+            _parameter: is_parameter,
+            ..Default::default()
+        };
+        return out.push_on(state);
+    }
+
+    fn push_on(mut self, state: &mut State) -> usize {
+        let idx = state.nodes.len();
+        self._idx = idx;
+        state.nodes.push(self);
+        return idx;
+    }
+}
+
+pub fn checkpoint(node: usize, name: &str, state: &mut State) -> usize {
+    let out = Node {
+        data: state[node].data,
+        name: format!("Checkpoint({})", name),
+        _prev: (Some(node), None),
+        _op: Op::Pass,
+        ..Default::default()
+    };
+    out.push_on(state)
+}
```

Comment:
Checkpoint is a simple pass-through node that allows us to break back propagation at critical places.